### PR TITLE
Introduce versioning for SDKs

### PR DESCRIPTION
**Pull request by brunoocasali**
The idea here is to help the users understand how the SDKs will deal with versioning now that we are entering the v1 era.

For reviewers:

I'm not sure about the way I wrote this initial version. I spilled everything from my mind 😄. Please let me know if there is a proper way of doing it.
The initial part is very similar to what Clem wrote in the [engine](https://github.com/meilisearch/engine-team/blob/main/resources/versioning-policy.md) side.

@dichotommy, this is my answer to the question you raised [here](https://github.com/meilisearch/documentation/issues/2113) and during our private conversation about the subject. Let me know if this can help you.

@meilisearch/integration-team feel free to think outside the box here and suggest improvements

- [ ] Create issues regarding the addition of a compatibility version table.

### CODE REVIEW

File: resources/versioning.md
Comment (bidoubiwa) [inline resources/versioning.md:65]:
```suggestion
And also if we get a Meilisearch v2 in the future, there is no guarantee that we will have Meilisearch Dart v2 as well.
```

File: resources/versioning.md
Comment (bidoubiwa) [inline resources/versioning.md:72]: `or` or `and` ?
```suggestion
| v1.0 | >= v0.15.0 |
```

File: resources/versioning.md
Comment (bidoubiwa) [inline resources/versioning.md:75]:
```suggestion
Given the example above, it means that the integration in v0.15 works with Meilisearch v1.1, but not all the new features introduced in the v1.1 will work with this integration version.
```

File: resources/versioning.md
Comment (bidoubiwa) [inline resources/versioning.md:78]:
```suggestion
## When to we update SDKs to v1?
```

File: resources/versioning.md
Comment (bidoubiwa) [inline resources/versioning.md:81]: > Or external facts that we may not control

For example?

Comment (bidoubiwa) [inline resources/versioning.md:81]:
```suggestion
Or external facts that we may not control. If you only want to use a v1 version of a SDKs please, open an issue in the integration-guides. It helps us decide when to update to v1. 
```

File: resources/versioning.md
Comment (bidoubiwa) [inline resources/versioning.md:88]:
```suggestion
:warning: Most of the SDKs have methods like get `GetAllRawIndexesAsync`, `allRaw`, `rawSearch` and similar. These methods have NO guarantee at all, since you're accessing directly the response given by the Meilisearch engine.
```

> have NO guarantee at all

Guarantee of what?

File: resources/versioning.md
Comment (bidoubiwa) [inline resources/versioning.md:89]:
```suggestion
We suggest avoiding using them.
```

**Review by curquiza — COMMENTED** [review]
Thanks for this guide Bruno! Really useful

I would add some warning
- it's a versioning policy, so general rules you want to follow with your team -> I have sometimes the feeling here we are justifying more than explaining what we do. Justifying the rules is good as well. But it should not be the center.
- Give more place to the community to express themself. We say we need feedback to make SDK reach v1, so we need to encourage people to help us on this! 😄

File: resources/versioning.md
Comment (brunoocasali) [reply resources/versioning.md:88]: I'll rework in the entire phrase, thanks for the highlight!

File: resources/versioning.md
Comment (brunoocasali) [reply resources/versioning.md:18]: I've changed the wording :)

File: resources/versioning.md
Comment (brunoocasali) [reply resources/versioning.md:53]: On the JS side, for example, it is very common to mention this change as a breaking:

- https://github.com/ember-cli/ember-cli-htmlbars/releases/tag/v6.0.0
- https://github.com/mochajs/mocha/releases/tag/v10.0.0
- https://github.com/typed-ember/ember-cli-typescript/releases/tag/v5.0.0
- https://github.com/emberjs/ember-qunit/releases/tag/v6.0.0

File: resources/versioning.md
Comment (brunoocasali) [reply resources/versioning.md:65]: After Charlotte's suggestions, I changed to:

> It means there is no guarantee that since Meilisearch engine is v1 we will have: Meilisearch Dart v1, Meilisearch JavaScript v1, Meilisearch .NET v1 and so on...
And also if we get a Meilisearch v2 in the future, there is no guarantee that we will have Meilisearch Dart v2 as well.

But yeah, it is just to inform the users, that they should not expect that.

File: resources/versioning.md
Comment (brunoocasali) [reply resources/versioning.md:57]: That's good you approached this topic because this is one of my doubts.

I would like to have this "justification and communication" written somewhere, and also rules to be followed. Where do you think they should go (the justification and communication)?

File: resources/versioning.md
Comment (brunoocasali) [reply resources/versioning.md:88]: This is more a communication to the users about those methods, I would like to mention they are not "protected" by the layer of the SDK. My new version after Charlotte's comments is: WDYT?

----
:warning: Most of the SDKs have methods like get `GetAllRawIndexesAsync`, `allRaw`, `rawSearch` and similar.

These methods made the request directly to the Meilisearch engine and they also respond with the raw answer.
It means you should take care of possible inconsistencies that may appear in the Meilisearch engine response.

So, if possible don't use them.

File: resources/versioning.md
Comment (curquiza) [reply resources/versioning.md:53]: It's as you wish then! 👌

File: resources/versioning.md
Comment (curquiza) [reply resources/versioning.md:57]: For me it's more about the form. You start by saying, "there is no plan to make SDK into v1"

This is not what the users want to know when they read your guide. They want to know 
- if the SDK versions will follow the version of Meilisearch or not 
- and then, which version of SDK is compatible with which version of Meilisearch, and how they can get the information

I would start to give your rules first, and then, the justifications, following by "we are open to feedback and your needs...blabla". Not the contrary

File: resources/versioning.md
Comment (curquiza) [reply resources/versioning.md:88]: Hum I understand better, it's to do a warning saying "since these methods are only based on the Meilisearch response, ensure you use the expected version of Meilisearch".
But for me this is not a warning only for these kinds of methods. The warning is important but should include more cases. Indeed, a user trying to use a new feature, let's say "multi-index-search", by calling the `muli_index_search()` method but with the wrong version of Meilsearch, will have the same issue. The user should take care in general about using the right version of Meilisearch with the right SDK version.

> So, if possible don't use them.

I don't understand why, since Meilisearch will not break

Comment (bidoubiwa) [reply resources/versioning.md:88]: I think that the difference with the other methods is that we cannot introduce backward compatibility if we'd like to.

Comment (curquiza) [reply resources/versioning.md:88]: Ok, then the guide needs to be more precise, because it's not obvious 😇 
however since we don't even have planned v2, I'm not sure this is an issue we should care of now.

File: resources/versioning.md
Comment (brunoocasali) [reply resources/versioning.md]: Updated!

**Review by bidoubiwa — CHANGES_REQUESTED** [review]

File: resources/versioning.md
Comment (brunoocasali) [reply resources/versioning.md:3]: Good catch!

File: resources/versioning.md
Comment (brunoocasali) [reply resources/versioning.md]: I added some content about it, maybe it now better explained.

File: resources/versioning.md
Comment (brunoocasali) [reply resources/versioning.md]: Fixed!

**Review by bidoubiwa — DISMISSED** [review]
LGTM 🔥

File: resources/versioning.md
Comment (alallema) [inline resources/versioning.md:53]: Does that mean we will have to update it each time we drop the support of a version?

File: resources/versioning.md
Comment (alallema) [inline resources/versioning.md:90]: Is there still some `all` like `getAll` somewhere in some SDK?

**Review by alallema — COMMENTED** [review]
Great ✨ ✨✨!!!
Just few questions

File: resources/versioning.md
Comment (brunoocasali) [reply resources/versioning.md:53]: Yeah! But this kind of change does not happen in a high-frequency, may once a year or more?

File: resources/versioning.md
Comment (brunoocasali) [reply resources/versioning.md:90]: I would have to check that, but they just have a bad naming they are not so harmful like the `raw` ones.

**Review by alallema — APPROVED** [review]

File: resources/versioning.md
Comment (brunoocasali) [reply resources/versioning.md:65]: Cool, good to know, I was scared that what I said earlier was not comprehensible :)

File: resources/versioning.md
Comment (brunoocasali) [reply resources/versioning.md:75]: I changed from:

> Given the example above, it means that the integration in v0.15 works with Meilisearch v1.1, but may only some of the new features introduced in v1.1 will work with this integration version.

to: 

> Given the table above, a user using Meilisearch v1.1 does not need to have the v0.18 version of its SDK.
Of course, some features will not be present, but they should still work.

WDYT?

Comment (curquiza) [reply resources/versioning.md]: Are you sure you need to talk about this? Starting from Meilisearch v1.0, every SDK will be compatible with any version of Meilisearch >= v1.0
I'm not sure the users would understand that "it is compatible, but some features would be missing". From my point of view, if some features are missing, it's not really compatible

Comment (curquiza) [reply resources/versioning.md]: It's part of the confusion I put [here](https://github.com/meilisearch/integration-guides/issues/237#issuecomment-1422986421)
You need to define what is "compatible" for you

Comment (brunoocasali) [reply resources/versioning.md]: I left a comment [here](https://github.com/meilisearch/integration-guides/issues/237#issuecomment-1431882992) where I expect we will agree with the term.
 
So, since the versioning table is not helpful, I'll remove it from this guide. And in the future, we can work on a better version where we can generate it using the code samples.

File: resources/versioning.md
Comment (curquiza) [inline resources/versioning.md:30]:
```suggestion
There is no strict policy yet regarding the minimum support of the language version, but we have still defined some rules. Example:
```

File: resources/versioning.md
Comment (curquiza) [inline resources/versioning.md:36]: You can turn it differently to make it more user-oriented, but still clear for the integration team, something like "we always try to avoid dropping support because we know we are aware some users could be left behind"

**Review by curquiza — CHANGES_REQUESTED** [review]
Sorry Burno, another little change I've just noticed (I should have seen it in my previous review). 

Otherwise it's perfect. Not too much information. Just what you and your team need to start. Feel free to add anything in the future, depending on the user confusion you will spot 😄

File: resources/versioning.md
Comment (brunoocasali) [reply resources/versioning.md:30]: Since we didn't use the first person tense, I'll change to: "There is no strict policy yet regarding the minimum support of the language version, but some ground rules were defined. Example:"

File: resources/versioning.md
Comment (brunoocasali) [reply resources/versioning.md]: Good idea, done! I just changed the message itself to make it in third person!

File: resources/versioning.md
Comment (curquiza) [reply resources/versioning.md]: Better indeed!

File: resources/versioning.md
Comment (curquiza) [inline resources/versioning.md:26]: I'm not sure I really understand the title
Plus there is one `##` title, and only one `###` title inside. Not sure the `##` title is needed.

File: resources/versioning.md
Comment (curquiza) [inline resources/versioning.md:28]: Why `new`?

**Review by curquiza — COMMENTED** [review]
I don't want to block this PR anymore, so this is my final comments, the rest of the job is good to me 👌 
Don't need my approval to merge this PR 😊 

Thanks @brunoocasali for the job, this guide is useful for the community! 😄

File: resources/versioning.md
Comment (bidoubiwa) [inline resources/versioning.md:73]: It is indeed a bit confusing for the user. It does look like the user requires v0.18.0. 
Additionally, does it mean that, for example, javascript does not have a valid version for v1.0 as it is missing some features like the CSV document addition?

Maybe we should remove this table and avoid unnecessary complexity

File: resources/versioning.md
Comment (brunoocasali) [reply resources/versioning.md:26]: You're right, I'll remove the H2 title :)

File: resources/versioning.md
Comment (brunoocasali) [reply resources/versioning.md:28]: What I wanted to mean with this were to imply that when we want to adopt a "new" runtime (different than the current supported ones), we should release a major (if we do the cut as explained). Maybe a better choice could be: `Minimum supported runtime`

**Pull request merged**
